v2.0.0 Bug Fixes and resolved tech debt

50ap5ud5 opened the pull request:
### Bug Fixes
- ShellBaseBlockEntity always having the wrong TARDIS_ID being set during exterior shell placement in TardisExteriorManager. This breaks most functions of the Tardis, such as teleporting players to the wrong dimension, and breaking the Tardis Key's ability to unlock/lock the Tardis
    - Added check to ``TardisHelper.teleportEntityTardis`` to prevent entities being teleported to an invalid dimension if trying to enter via an exterior shell which doesn't have a valid TARDIS_ID due to above bug.
    - Since players won't be able to enter their Tardises anymore (better this than wrong teleports), they can "fix" it by teleporting into the Tardis Interior via `/tardis_refined interior` and forcing a block update on the Shell block, such as changing exteriors or flying the Tardis to a different location.
    - On multiplayer servers, administrator intervention may be required.
- Added ``TardisExteriorManager#setOrUpdateExteriorBlock`` as a common logic for placing Exterior blocks whilst preserving critical data such as ShellTheme and Patterns
- Replaced ``TardisExteriorManager#placeExteriorBlock`` with ``startLanding`` and ``setOrUpdateExteriorBlock``, where ``setOrUpdateExteriorBlock`` executes the block placement, and ``startLanding`` updates Tardis data and chunkloads the area.
- Enhanced ``TardisExteriorManager#triggerShellRegenState`` with a boolean flag to unify logic for updating the REGEN blockstate property on the Shell Block when the Tardis is reconfiguring its Desktop.
- KeyItem using wrong logic to append strings to translatable tooltip strings. Fixes scenario where tooltip displayed ``*message.tardis_refined.tooltip_in_flight*``
- Fixed ``TardisExteriorManager.setDoorClosed`` using confusing logic when trying to determine if we are trying to close the door or not. Previously, it was using the ``locked`` field to determine "closed" state, which was erroneous.
- Large amount of duplicate control entities present on Nuka Console
- Fixed placement of monitor control for all consoles. Fixes some of the behaviour described in #283. **To be tested**
- Added dimension load fix from #296 
- Missing translations for all default Hum sounds (used one translation key to cover all scenarios- "TARDIS hums")
- Missing translation for SCREWDRIVER_CONNECT sound
- Missing translation for LOW_FUEL sound
- Fixed FlightDance control re-alignment logic in ControlEntity allowing player to activate the control. Fixes Issue #269
     - **Example**: "Randomiser" control is considered "misaligned" by FlightDanceManager.
     - When player clicks on the Randomiser control, the control is realigned BUT the Tardis coordinates are ALSO randomised.
     - This causes the flight destination to be changed and the flight duration to be recalculated, which was unintended.
- Fixed Gravity Shaft causing player to open elytra when JUMP key is pressed.
     - NOTE: Due to the patch using ``Player#onGround`` this means all game behaviours relating to when the player is on the ground (hunger, damage etc.) will also be applied
- Fixed Tardis Flight Loop not playing (Combination of undocumented logic being called in multiple call sites, and the volume being set to zero causing the SoundManager to skip playing it). Fixes Issue #293
- Fixed several sound entries having the incorrect registry names, which caused any data-generated translations for their subtitles to always be incorrect.
- `HumSelectionScreen` using erroneous logic for parsing Hum text components, causing log spam each time the screen is opened.

### Changes
- Refactor of LoopingSound and TardisClientLogic to unify logic, remove duplicate call sites, document logic.
    - Moved SoundInstance registry entries into their own class, cleaning up method names
- Made TardisPilotingManager and TardisExteriorManager explicitly update TardisClientData in a few critical methods such as ``endFlight``, to ensure that sounds dependent on that data is played at the correct time.

### Additions
- Added a boolean flag to the `Control` object to allow for controls to be excluded from the list of possible controls to use for the FlightDanceManager. This is flexible enough to allow for addon mods to modify and customise behaviour.

## Notes
- The way we generate control positions needs to be improved. The output generated by default is off by a few decimal places that causes the control to be placed sometimes quite far away from its intended location.
- Need to be careful of generating duplicate controls. They cause performance hits and prevent the player from completing Flight Dance events due to multiple controls placed in the same location with the same hitbox size.
- TardisClientLogic needed a cleanup. The LoopingSounds logic was making duplicate call sites using undocumented logic.
- As a general convention: SoundEvent IDs need to be the same as their field names. This way we can be sure that the subtitle translations are generated correctly via data-gen.

50ap5ud5 committed:
Fixed desktop reconfig losing pattern data on shell, other fixes

-ShellBaseBlockEntity always having the wrong TARDIS_ID being set during exterior shell placement in TardisExteriorManager. This breaks most functions of the Tardis, such as teleporting players to the wrong dimension, and breaking the Tardis Key's ability to unlock/lock the Tardis
-Added check to TardisHelper.teleportEntityTardis to prevent entities being teleported to an invalid dimension if trying to enter via an exterior shell which doesn't have a valid TARDIS_ID due to above bug.
-KeyItem using wrong logic to append strings to translatable tooltip strings.
-TardisExteriorManager.setDoorClosed using confusing logic when trying to determine if we are trying to close the door or not.
-Temp patch for in-flight sound not playing due to sound engine being spammed with too many requests

50ap5ud5 committed:
Added missing subtitle translations for Tardis hums, interacting with misaligned controls will no longer activate the control's function

-Fixes players being able to unintentionally activate the control.
-Fixed HumSelectionScreen not parsing names correctly, causing incessant log spam

50ap5ud5 committed:
Fix control placement of large controls for all consoles

50ap5ud5 committed:
Fixed huge amount of duplicate controls on Nuka, cleaned up VictorianConsoleTheme, fixed Myst

Jeryn99 committed:
Fixes Dodgy Dimensions and add Create Compat

50ap5ud5 committed:
Control adjustments

50ap5ud5 committed:
adjustments

50ap5ud5 committed:
Fix Gravity shaft causing player to open Elytra when jumping.

-Stops player flying whilst in gravity shaft and taking damage
-Side effect: Hunger and other behaviours that occur when player is touching the ground will apply.

50ap5ud5 committed:
Update mod version, wiki, issue template, credits

50ap5ud5 committed:
Fixed and refactored looping sound implementation

-Moved all AbstractTickableSoundInstance registry entries to their own class
-Unified logic for all looping sound instances.
-Massively cleaned up TardisClientLogic
-Fixed particle playing using inconsistent methods
-Added overloads to Client
.playParticle

50ap5ud5 committed:
Fixed Hum sound registry names causing subtitle to be incorrect

50ap5ud5 committed:
Documented and refactored HumEntry method names for clarity

50ap5ud5 committed:
Finally fixed LoopingSound not playing correctly.

-Added more specific implementations of LoopingSound to allow for more structured logic to be defined

50ap5ud5 committed:
Fix FlightDance not actually excluding non-essential controls, flight loop sounds not playing when needed due to client data not being synced

-Fixed flight loop sound always playing even when flight has ended.
-Fixed typo in GlobalConsoleBlockEntity.getControlEntityList

Jeryn99 commented on common/src/main/java/whocraft/tardis_refined/common/GravityUtil.java:
```diff
@@ -18,7 +18,7 @@ public class GravityUtil {
     private static final double ACCELERATION = 0.2;
     public static boolean isInAntiGrav(Player playerBox, AABB box, Level level) {
-        if(level.dimensionTypeId() != TRDimensionTypes.TARDIS) return false;
+//        if(level.dimensionTypeId() != TRDimensionTypes.TARDIS) return false; //Allow sound to play no matter what dimension the block is placed in
```
Is this commented code to remain or to be removed?

Jeryn99 replied:
Could also be a tag?

50ap5ud5 replied:
Commented because I was going to ask if it was ok to do so. 

Previously the playing of the sound was restricted to the Tardis dimension.
However, players can craft this block so will be using it outside the Tardis dimension.
For a consistent experience I think we should enable the sound to be played anywhere, so this line of code should be removed.

Not sure how to use a dimension tag, didn't know that was possible.

Jeryn99 commented on common/src/main/java/whocraft/tardis_refined/common/tardis/manager/FlightDanceManager.java:
```diff
@@ -33,8 +33,16 @@ public boolean isDancing() {
+        var allControlsOnConsole = controllerConsole.getControlEntityList();
+        List<ControlEntity> updatedList = new ArrayList<>(allControlsOnConsole); //Copy over all entries to a new which we can manipulate
+        for(ControlEntity controlEntity : allControlsOnConsole){ //Out of all controls in the original control list, remove any from our copy
+            if(controlEntity.controlSpecification().control() == TRControlRegistry.THROTTLE.get()
```
We could perhaps have a boolean on controls for this? If addons decided to add controls they would need to mixin this and we should avoid that if we can

50ap5ud5 replied:
Yep agreed, we can add a boolean field on the Control class. Would it be only used for exclusion from the Flight Dance?

50ap5ud5 replied:
Added, kept it generically named.

Jeryn99 replied:
@50ap5ud5 this is where I meant to say the tag thing

Jeryn99 reviewed (approved):
A much needed climb out of technical debt, love the approaches used and annoyed that we didn't approach it from this angle to begin with


Well done 50

50ap5ud5 committed:
Fix FlightSound not resetting its volume when needed

-Needs to be a non-zero value when we need to play it. When not in flight the volume set to zero so that we don't play it, we need to make sure we do reset this value

50ap5ud5 committed:
Added a flag to Control to exclude it from the list of controls to use for FlightDanceManager
